From f624c0347c9e9e94d88ca85caa07ab945969be0c Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Mon, 1 Dec 2008 21:28:31 +0000
Subject: [PATCH] CONFIGURATION-321: The iterator returned by
 HierachicalConfiguration.getKeys(String prefix) now contains the prefix if
 this key is contained in the configuration.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@722238 13f79535-47bb-0310-9956-ffa450edef68
---
 .../HierarchicalConfiguration.java            |  6 ++++
 .../TestHierarchicalConfiguration.java        | 29 +++++++++++++++++++
 xdocs/changes.xml                             |  5 ++++
 3 files changed, 40 insertions(+)

diff --git a/src/java/org/apache/commons/configuration/HierarchicalConfiguration.java b/src/java/org/apache/commons/configuration/HierarchicalConfiguration.java
index 1c660c979b..81e42e5dd6 100644
--- a/src/java/org/apache/commons/configuration/HierarchicalConfiguration.java
+++ b/src/java/org/apache/commons/configuration/HierarchicalConfiguration.java
@@ -824,6 +824,12 @@ public Iterator getKeys()
     public Iterator getKeys(String prefix)
     {
         DefinedKeysVisitor visitor = new DefinedKeysVisitor(prefix);
+        if (containsKey(prefix))
+        {
+            // explicitly add the prefix
+            visitor.getKeyList().add(prefix);
+        }
+
         List nodes = fetchNodeList(prefix);
 
         for (Iterator itNodes = nodes.iterator(); itNodes.hasNext();)
diff --git a/src/test/org/apache/commons/configuration/TestHierarchicalConfiguration.java b/src/test/org/apache/commons/configuration/TestHierarchicalConfiguration.java
index 9eddd882a1..c4b0a62ca1 100644
--- a/src/test/org/apache/commons/configuration/TestHierarchicalConfiguration.java
+++ b/src/test/org/apache/commons/configuration/TestHierarchicalConfiguration.java
@@ -357,6 +357,35 @@ public void testGetKeysString()
                 new String[] {"url", "user" });
     }
 
+    /**
+     * Tests getKeys() with a prefix when the prefix matches exactly a key.
+     */
+    public void testGetKeysWithKeyAsPrefix()
+    {
+        config.addProperty("order.key1", "value1");
+        config.addProperty("order.key2", "value2");
+        Iterator it = config.getKeys("order.key1");
+        assertTrue("no key found", it.hasNext());
+        assertEquals("1st key", "order.key1", it.next());
+        assertFalse("more keys than expected", it.hasNext());
+    }
+
+    /**
+     * Tests getKeys() with a prefix when the prefix matches exactly a key, and
+     * there are multiple keys starting with this prefix.
+     */
+    public void testGetKeysWithKeyAsPrefixMultiple()
+    {
+        config.addProperty("order.key1", "value1");
+        config.addProperty("order.key1.test", "value2");
+        config.addProperty("order.key1.test.complex", "value2");
+        Iterator it = config.getKeys("order.key1");
+        assertEquals("Wrong key 1", "order.key1", it.next());
+        assertEquals("Wrong key 2", "order.key1.test", it.next());
+        assertEquals("Wrong key 3", "order.key1.test.complex", it.next());
+        assertFalse("More keys than expected", it.hasNext());
+    }
+
     public void testAddProperty()
     {
         config.addProperty("tables.table(0).fields.field(-1).name", "phone");
diff --git a/xdocs/changes.xml b/xdocs/changes.xml
index 0c621f201f..0671294fee 100644
--- a/xdocs/changes.xml
+++ b/xdocs/changes.xml
@@ -157,6 +157,11 @@
         ConfigurationDynaBean now works properly with indexed properties
         stored internally in the underlying configuration as arrays.
       </action>
+      <action dev="oheger" type="fix" issue="CONFIGURATION-321">
+        The iterator returned by HierarchicalConfiguration.getKeys(String prefix)
+        now also contains the prefix if this key is contained in the
+        configuration.
+      </action>
       <action dev="ebourg" type="fix" issue="CONFIGURATION-320">
         XMLPropertyListConfiguration is no longer limited to 32 bits integers.
       </action>
